File: Python/53.py
def maxSubArray(nums) -> int:
    length = len(nums)
    res = temp = nums[0]
    for i in range(1, length):
        temp = max(temp + nums[i], nums[i])
        if temp > res:
            res = temp
    return res
nums = [-2,1,-3,4,-1,2,1,-5,4]
print(maxSubArray(nums))

File: Java/53.java
class Solution53 {
    public int maxSubArray(int[] nums) {
        int[] dp = new int[nums.length]; // dp[i]表示以nums[i]结尾的最大和
        dp[0] = nums[0]; // 边界
        int res = dp[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] + dp[i - 1] > nums[i])
                dp[i] = nums[i] + dp[i - 1];
            else
                dp[i] = nums[i];
            res = Math.max(res, dp[i]);
        }
        return res;
    }
}